### PR TITLE
Initial typing: add mypy and fix immediate typing errors

### PR DESCRIPTION
**Pull request by arusahni**
Start down the path for #137 

I have _a lot_ of diffs adding typing on my local branch. The goal here is to get some of the baseline choices discussed first (e.g., `mypy` vs `pyright`) and not land too many structural changes.

### Included

1. Drop Python 2.7 from `tox`. Since we're committing to annotations and it's well past EOL I think this is noncontroversial.
2. Use `mypy`. I'm ambivalent here - I've been on teams who've used both. `mypy` is the official option, while `pyright` is closer to what folks may get out of their IDEs.
3. Some selective ignores. Do not typecheck the Config constructor because it's pretty hairy. I think this is a prime candidate for implementation tweaking (especially since it does some stuff I'm not fully understanding the purpose of, such as using `eval` to inspect the globals to form an API key).
4. Use Poetry for both library deps and development deps. This includes a `pyproject.toml` file, as well as the `poetry.lock` file. Poetry themselves are ambivalent about libraries versioning the file, so I'm open to excluding it.

Once we're happy with the shape of the above, I can start creating PRs with my unstaged changes.

### CODE REVIEW

**Comment by arusahni** [comment]
I've pushed a commit that switches to Poetry.

**Comment by 2e0byo** [comment]
sorry for the delay; I've been sick. I'll get on the rest of this this week, thanks!

**Comment by arusahni** [comment]
@2e0byo I wanted to check in (just in case this got lost in your queue). I hope you're feeling better! If not, then well-wishes :pray:

**Comment by 2e0byo** [comment]
Much better thank you! Pretty sick for a few weeks but all over.  I'll try to get on this tomorrow

**Comment by 2e0byo** [comment]
Sorry for the massive delay on this.  @tamland this PR is a good starting point for typing (I think we can be a bit more implicit inside some functions, but thanks @arusahni for the hard (and tedious) work of getting the whole thing passing).  Config needs rewriting.  mypy passes.

When I started testing I realised the tests need work.  This PR breaks python < 3.10 (`TypeAlias`, subscripts on `list` and `dict`, etc) and .  We can fix that, but what version *should* we support?  (@arusahni  if it wasn't for this I'd just merge).  Regardless I think we should merge asap, get this in, and then I'll open a cleanup PR.  Some tests need fixing, some things can be tidied up, we can record (sanitised) network data with `vcr` and run in CI, we should lint everything, etc. 

tl;dr @tamland  (@BlackLight) if you're happy with breaking python <3.10: merge.  Else let's define a minimal version (3.7/8 should be dropped IMHO).

*<sup><sub>If we are going with 3.10 those `Optional`s are *going*... but that's a trivial change<sub><sup>

**Comment by 2e0byo** [comment]
@tehkillerbee  tagging as I don't think the updated comment will trigger an event.  @tamland sorry to spam you

**Comment by tehkillerbee** [comment]
@2e0byo yep, I get most notifications. Anyways, I think 3.7 is too old and 3.8 is EOL next year if I recall. I would like to support 3.9 and above, although I'm not sure if anyone is still using tidalapi on such old versions of python. Perhaps on embedded platforms? RPi or similar?

**Comment by 2e0byo** [comment]
Right, I'll tweak for 3.9 for now and merge; I've blocked this for long enough.  Will make a long term PR as well.

**Comment by tehkillerbee** [comment]
> Right, I'll tweak for 3.9 for now and merge; I've blocked this for long enough. Will make a long term PR as well.

Great to hear. Then we can get PyPi updated with the latest fixes soon.

**Comment by 2e0byo** [comment]
@arusahni made a few tweaks, sorry to push to your master (you can revert with a force push if you want to drop it; was the quickest way to get the changes in whilst leaving your name attached).  Thanks for the hard work and sorry for the delay!  Let's discuss the remaining once the next merge lands---it'll probably make a lot of conflicts.

**Comment by arusahni** [comment]
@2e0byo No worries, and thanks for getting this over the line! I'm no stranger to weird rebases, so I can see about getting the rest queued up tomorrow.
